src/sentry/web/frontend/auth_login.py

```diff
@@ -67,11 +69,168 @@ def run_callbacks(self, request: Request):
+        """
+        if is_valid_redirect(url=next_uri, allowed_hosts=(request.get_host(),)):
+            return self.redirect(url=next_uri)
+        return self.redirect_to_org(request=request)
+
+    def redirect_to_org_specific_login(self) -> HttpResponseRedirect:
+        """
+        If Sentry is in single org mode this redirects users to their specific handler.
+        """
+        org = Organization.get_default()
+        next_uri = reverse("sentry-auth-organization", args=[org.slug])
+        return HttpResponseRedirect(redirect_to=next_uri)
+
+    def should_redirect_to_sso_login(self, request: Request) -> bool:
+        """
+        Checks if a GET request to our login page should redirect to SSO login.
+        """
+        non_sso_urls = [
+            reverse("sentry-auth-organization", args=[request.subdomain]),
+            reverse("sentry-register"),
+        ]
+        return (
+            request.subdomain
+            and self.org_exists(request=request)
+            and request.path_info not in non_sso_urls
+        )
+
+    def get_sso_redirect(self, request: Request) -> HttpResponseRedirect:
+        """
+        Returns a redirect response that will take a user to SSO login.
+        """
+        url_prefix = generate_organization_url(org_slug=request.subdomain)
+        url = absolute_uri(
+            url=reverse("sentry-auth-organization", args=[request.subdomain]), url_prefix=url_prefix
+        )
+        if request.GET:
+            url = f"{url}?{request.GET.urlencode()}"
+        return HttpResponseRedirect(redirect_to=url)
+
+    def get_registration_page(self, request: Request, context: dict, **kwargs) -> HttpResponse:
+        """
+        Returns the standard registration page.
+        """
+        register_form = self.get_register_form(
+            request=request, initial={"username": request.session.get("invite_email", "")}
+        )
+        context.update(
+            {
+                "op": "register",
+                "CAN_REGISTER": True,
+                "register_form": register_form,
+            }
+        )
+        return self.respond_login(request=request, context=context, **kwargs)
+
+    def get_login_page(self, request: Request, context: dict, **kwargs) -> HttpResponse:
+        """
+        Returns the standard login page.
+        """
+        op = "sso" if request.GET.get("op") == "sso" else "login"
+        login_form = AuthenticationForm(request=request)
+        context.update(
+            {
+                "op": op,
+                "login_form": login_form,
+            }
+        )
+        return self.respond_login(request=request, context=context, **kwargs)
+
+    def post(self, request: Request, **kwargs) -> HttpResponse:
+        op = request.POST.get("op")
+        if op == "sso" and request.POST.get("organization"):
+            return self.redirect_post_to_sso(request=request)
+
+        default_context = self.get_default_context(request=request, **kwargs)
```

Comment:
`default_context = self.get_default_context(request=request)` can be moved into the functions `self.handle_register_form_submit()` and `self.handle_login_form_submit()`.